File: 04_interest_rate_swap_pricing/python/swap.py
from fixed_leg import FixedLeg
from floating_leg import FloatingLeg


class InterestRateSwap:
    """
    Plain vanilla fixed-for-floating IRS.

    Payer swap:
    - pay fixed
    - receive floating

    Receiver swap:
    - receive fixed
    - pay floating
    """

    def __init__(
        self,
        notional,
        fixed_rate,
        maturity,
        payment_frequency=2,
        payer=True
    ):

        self.notional = notional
        self.fixed_rate = fixed_rate
        self.maturity = maturity
        self.payment_frequency = payment_frequency
        self.payer = payer

        step = 1 / payment_frequency

        self.payment_dates = []

        t = step

        while t <= maturity + 1e-10:

            self.payment_dates.append(round(t, 10))

            t += step

        self.fixed_leg = FixedLeg(
            notional,
            fixed_rate,
            self.payment_dates
        )

        self.floating_leg = FloatingLeg(
            notional,
            self.payment_dates
        )

    def npv(self, curve):
        """
        Swap NPV.

        Payer:
        float - fixed

        Receiver:
        fixed - float
        """

        fixed_pv = self.fixed_leg.pv(curve)

        float_pv = self.floating_leg.pv(curve)

        if self.payer:

            return float_pv - fixed_pv

        else:

            return fixed_pv - float_pv

    def par_swap_rate(self, curve):
        """
        Compute par swap rate.

        Par rate makes swap NPV = 0.
        """

        numerator = (
            1 - curve.discount_factor(self.maturity)
        )

        denominator = 0.0

        previous = 0.0

        for t in self.payment_dates:

            accrual = t - previous

            denominator += (
                accrual
                * curve.discount_factor(t)
            )

            previous = t

        return numerator / denominator

File: 04_interest_rate_swap_pricing/cpp/swap.cpp
#include "swap.hpp"
#include <iostream>

InterestRateSwap::InterestRateSwap(
    double N,
    double K,
    double T
){

    notional = N;
    fixed_rate = K;
    maturity = T;

    // ------------------------------------------------
    // Generate semiannual payment schedule
    // ------------------------------------------------

    int frequency = 2;

    int n_payments =
        static_cast<int>(
            maturity * frequency
        );

    for(int i=1;
        i<=n_payments;
        ++i){

        double t =
            static_cast<double>(i)
            / frequency;

        payment_dates.push_back(t);
    }
}

double InterestRateSwap::npv(
    YieldCurve& curve
){

    double fixed_pv = 0.0;
    double float_pv = 0.0;

    double previous = 0.0;

    for(double t : payment_dates){

        double accrual =
            t - previous;

        double fwd =
            curve.forward_rate(
                previous,
                t
            );

        double df =
            curve.discount_factor(t);

        fixed_pv +=
            notional
            * fixed_rate
            * accrual
            * df;

        float_pv +=
            notional
            * fwd
            * accrual
            * df;

        previous = t;
    }

    std::cout
        << "Final Fixed Leg PV = "
        << fixed_pv
        << std::endl;

    std::cout
        << "Final Floating Leg PV = "
        << float_pv
        << std::endl;

    return float_pv - fixed_pv;
}
